pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java

```diff
@@ -221,18 +256,30 @@ protected void doAddRecord(MutableSegment segment, RecordInfo recordInfo) {
   protected GenericRow doUpdateRecord(GenericRow record, RecordInfo recordInfo) {
     assert _partialUpsertHandler != null;
     AtomicReference<GenericRow> previousRecordReference = new AtomicReference<>();
+    AtomicBoolean outOfOrder = new AtomicBoolean();
     RecordLocation currentRecordLocation = _primaryKeyToRecordLocationMap.computeIfPresent(
         HashUtils.hashPrimaryKey(recordInfo.getPrimaryKey(), _hashFunction), (pk, recordLocation) -> {
           if (recordInfo.getComparisonValue().compareTo(recordLocation.getComparisonValue()) >= 0) {
-            _reuse.clear();
-            previousRecordReference.set(recordLocation.getSegment().getRecord(recordLocation.getDocId(), _reuse));
+            IndexSegment currentSegment = recordLocation.getSegment();
+            int currentDocId = recordLocation.getDocId();
+            ThreadSafeMutableRoaringBitmap currentQueryableDocIds = currentSegment.getQueryableDocIds();
+            if (currentQueryableDocIds == null || currentQueryableDocIds.contains(currentDocId)) {
+              // if delete is not enabled or previous record not marked as deleted
+              _reuse.clear();
+              previousRecordReference.set(currentSegment.getRecord(currentDocId, _reuse));
+            }
+          } else {
+            outOfOrder.set(true);
           }
           return recordLocation;
         });
+    GenericRow previousRecord = previousRecordReference.get();
     if (currentRecordLocation != null) {
       // Existing primary key
-      GenericRow previousRecord = previousRecordReference.get();
-      if (previousRecord != null) {
+      if (!outOfOrder.get()) {
+        if (recordInfo.isDeleteRecord() || previousRecord == null) {
```

Comment:
We can simplify this check to `if (previousRecord == null)` after applying the check when reading the previous record